libvast/test/module.cpp

```diff
@@ -1171,22 +1298,87 @@ TEST(YAML Module - order independent parsing - type aliases) {
 }
 
 TEST(YAML Module - order independent parsing - list_type) {
+  auto declaration = record{{
+    "types",
+    record{
+      {"type1", record{{"list", "type2"}}},
+      {"type2", record{{"list", "type3"}}},
+      {"type3", record{{"type", "string"}}},
+    },
+  }};
+  auto result = unbox(to_module2(declaration));
+  auto expected_result = module_ng2{
+    .types
+    = {type{"type3", string_type{}},
+       type{"type2", list_type{type{"type3", string_type{}}}},
+       type{"type1", list_type{
+                       type{"type2", list_type{type{"type3", string_type{}}}},
+                     }}}};
+  CHECK_EQUAL(result, expected_result);
+}
+
+TEST(YAML Type - order indepenedent parsing - map_type) {
+  std::vector<type> known_types;
+  auto declaration = record{{
+    "types",
+    record{
+      {"map_type",
+       record{
+         {"map", record{{"key", "type1"}, {"value", "type2"}}},
+       }},
+      {"type1", record{{"type", "count"}}},
+      {"type2", record{{"type", "string"}}},
+    },
+  }};
+  auto result = unbox(to_module2(declaration));
+  auto expected_result
+    = module_ng2{.types = {
+                   type{"type1", count_type{}},
+                   type{"type2", string_type{}},
+                   type{"map_type", map_type{type{"type1", count_type{}},
+                                             type{"type2", string_type{}}}},
+                 }};
+  CHECK_EQUAL(result, expected_result);
+}
+
+TEST(YAML Type - order indepenedent parsing - record_type) {
```

Comment:
These three unit tests implement the order-independent parsing described in sc-33719